x-pack/solutions/observability/plugins/streams/server/routes/streams/processing/simulation_handler.ts

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { IScopedClusterClient } from '@kbn/core/server';
+import { isEmpty } from 'lodash';
+import { FieldDefinitionConfig, RecursiveRecord } from '@kbn/streams-schema';
+import { calculateObjectDiff, flattenObject } from '@kbn/object-utils';
+import { SimulationFailedError } from '../../../lib/streams/errors/simulation_failed_error';
+import { DetectedMappingFailureError } from '../../../lib/streams/errors/detected_mapping_failure_error';
+import { NonAdditiveProcessorError } from '../../../lib/streams/errors/non_additive_processor_error';
+import { formatToIngestProcessors } from '../../../lib/streams/helpers/processing';
+import { ProcessingSimulateParams } from './route';
+
+export const prepareSimulationBody = (params: ProcessingSimulateParams) => {
+  const { path, body } = params;
+  // eslint-disable-next-line @typescript-eslint/naming-convention
+  const { processing, documents, detected_fields } = body;
+
+  const processors = formatToIngestProcessors(processing);
+  const docs = documents.map((doc, id) => ({
+    _index: path.id,
+    _id: id.toString(),
+    _source: doc,
+  }));
+
+  const simulationBody: any = {
+    docs,
+    pipeline_substitutions: {
+      [`${path.id}@stream.processing`]: {
+        processors,
+      },
+    },
+  };
+
+  if (detected_fields) {
+    const properties = computeMappingProperties(detected_fields);
+    simulationBody.component_template_substitutions = {
+      [`${path.id}@stream.layer`]: {
+        template: {
+          mappings: {
+            properties,
+          },
+        },
+      },
+    };
+  }
+
+  return simulationBody;
+};
+
+// TODO: update type once Kibana updates to elasticsearch-js 8.17
+export const executeSimulation = async (
+  scopedClusterClient: IScopedClusterClient,
+  simulationBody: ReturnType<typeof prepareSimulationBody>
+): Promise<any> => {
+  try {
+    // TODO: We should be using scopedClusterClient.asCurrentUser.simulate.ingest() once Kibana updates to elasticsearch-js 8.17
+    return await scopedClusterClient.asCurrentUser.transport.request({
+      method: 'POST',
+      path: `_ingest/_simulate`,
+      body: simulationBody,
+    });
+  } catch (error) {
+    throw new SimulationFailedError(error);
```

Comment:
do we need to catch/rethrow error here, instead of just propagating? I'm not sure if it applies here, but I think it's a good guiding principle to delegate to ES (including error handling) as much as we can, and then handle in the UI